File: Algorithm/python_source/charpter2_reverselist/level2/topic2_4链表相加/add_two_numbers.py
# coding=utf-8

from python.unit3_linklist.part2_hot_problem.list_node import init_list, ListNode, print_linked_list


class AddTwoNumbers:
    # 方法1 使用栈实现
    def addTwoNumbers(self, l1, l2):
        st1 = []
        st2 = []
        while l1:
            st1.append(l1.val)
            l1 = l1.next
        while l2:
            st2.append(l2.val)
            l2 = l2.next
        carry = 0
        dummy = ListNode(0)
        while st1 or st2 or carry:
            adder1 = st1.pop() if st1 else 0
            adder2 = st2.pop() if st2 else 0
            sum = adder1 + adder2 + carry
            carry = 1 if sum >= 10 else 0
            sum = sum - 10 if sum >= 10 else sum
            cur = ListNode(sum)
            cur.next = dummy.next
            dummy.next = cur
        return dummy.next

    # 方法2 使用链表反转实现
    def reverseList(self, head):
        prev = None
        curr = head
        while curr:
            nextTemp = curr.next
            curr.next = prev
            prev = curr
            curr = nextTemp
        return prev

    def addTwoNumbersI(self, l1, l2):
        ans = ListNode(0, None)
        DUMMY_HEAD, res = ans, 0
        p1, p2 = l1, l2

        while p1 != None or p2 != None or res == 1:

            ans.next = ListNode(0, None)
            ans = ans.next

            if p1 != None and p2 != None:
                sums = p1.val + p2.val
                if sums + res < 10:
                    ans.val = sums + res
                    res = 0
                else:
                    ans.val = sums + res - 10
                    res = 1
                p1, p2 = p1.next, p2.next

            elif p1 == None and p2 != None:
                sums = p2.val
                if sums + res < 10:
                    ans.val = sums + res
                    res = 0
                else:
                    ans.val = sums + res - 10
                    res = 1
                p2 = p2.next

            elif p2 == None and p1 != None:
                sums = p1.val
                if sums + res < 10:
                    ans.val = sums + res
                    res = 0
                else:
                    ans.val = sums + res - 10
                    res = 1
                p1 = p1.next

            else:
                ans.val = res
                res = 0
        return DUMMY_HEAD.next

    # 调用入口
    def addTwoNumbers2(self, l1, l2):
        return self.reverseList(self.addTwoNumbersI(self.reverseList(l1), self.reverseList(l2)))


if __name__ == '__main__':
    # 先构造题目要求的链表,简单构造即可
    nums1 = [6, 1, 7]
    list1 = init_list(nums1)

    nums2 = [2, 9, 5]
    list2 = init_list(nums2)

    addTwoNumbers = AddTwoNumbers()
    list3 = addTwoNumbers.addTwoNumbers2(list1, list2)
    print_linked_list(list3)

File: Algorithm/Leetcode/Code_Materials/c_or_cpp/charpter2_reverselist/level2/topic2_4链表相加/add_two_numbers.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>
#include "../../list_node.cpp"

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode *st1 = NULL;
    struct ListNode *st2 = NULL;
    struct ListNode *curr1 = l1;
    struct ListNode *curr2 = l2;
    while (curr1 != NULL)
    {
        struct ListNode *new_node = (struct ListNode *)malloc(sizeof(struct ListNode));
        new_node->val = curr1->val;
        new_node->next = st1;
        st1 = new_node;
        curr1 = curr1->next;
    }
    while (curr2 != NULL)
    {
        struct ListNode *new_node = (struct ListNode *)malloc(sizeof(struct ListNode));
        new_node->val = curr2->val;
        new_node->next = st2;
        st2 = new_node;
        curr2 = curr2->next;
    }

    struct ListNode *new_head = (struct ListNode *)malloc(sizeof(struct ListNode));
    new_head->val = -1;
    new_head->next = NULL;
    int carry = 0;

    while (st1 != NULL || st2 != NULL || carry != 0)
    {
        struct ListNode *a = (struct ListNode *)malloc(sizeof(struct ListNode));
        struct ListNode *b = (struct ListNode *)malloc(sizeof(struct ListNode));
        if (st1 != NULL)
        {
            a = st1;
            st1 = st1->next;
        }
        else
        {
            a->val = 0;
            a->next = NULL;
        }
        if (st2 != NULL)
        {
            b = st2;
            st2 = st2->next;
        }
        else
        {
            b->val = 0;
            b->next = NULL;
        }

        int get_sum = a->val + b->val + carry;
        int ans = get_sum % 10;
        carry = get_sum / 10;

        struct ListNode *cur = (struct ListNode *)malloc(sizeof(struct ListNode));
        cur->val = ans;
        cur->next = new_head->next;
        new_head->next = cur;
    }

    struct ListNode *result = new_head->next;
    free(new_head);
    return result;
}

int main()
{
    struct ListNode *list1 = NULL;
    struct ListNode *list2 = NULL;

    int a[] = {6, 1, 7};
    list1 = initLink(a, 3);

    int b[] = {2,9,5};
    list2 = initLink(b, 3);

    struct ListNode *q = addTwoNumbers(list1,list2);
    printList(q);

    return 0;
}